sway/desktop/fx_renderer.c

```diff
@@ -76,6 +83,153 @@ static const float transforms[][9] = {
 	},
 };
 
+void scissor_output(struct wlr_output *wlr_output,
+		pixman_box32_t *rect) {
+	struct sway_output *output = wlr_output->data;
+	struct fx_renderer *renderer = output->renderer;
+	assert(renderer);
+
+	struct wlr_box box = {
+		.x = rect->x1,
+		.y = rect->y1,
+		.width = rect->x2 - rect->x1,
+		.height = rect->y2 - rect->y1,
+	};
+
+	int ow, oh;
+	wlr_output_transformed_resolution(wlr_output, &ow, &oh);
+
+	enum wl_output_transform transform =
+		wlr_output_transform_invert(wlr_output->transform);
+	wlr_box_transform(&box, &box, transform, ow, oh);
+
+	fx_renderer_scissor(&box);
+}
+
+static int get_blur_size(int blur_passes, int blur_radius) {
+	return pow(2, blur_passes) * blur_radius;
+}
+
+int fx_get_container_expanded_size(struct sway_container *con) {
+	bool shadow_enabled = config->shadow_enabled;
+	if (con) shadow_enabled = con->shadow_enabled;
+	int shadow_sigma = shadow_enabled ? config->shadow_blur_sigma : 0;
+
+	bool blur_enabled = config->blur_enabled;
+	if (con) blur_enabled = con->blur_enabled;
+	int blur_size =
+		blur_enabled ? get_blur_size(config->blur_passes, config->blur_radius) : 0;
+	// +1 as a margin of error
+	return MAX(shadow_sigma, blur_size) + 1;
+}
```

Comment:
this can live in `render.c`, and be renamed `get_container_expanded_size`

Reply:
We could also move get_blur_size to render.c if we can manage to move all of the damage tracking functions there

Reply:
Hmmm. The `fx_get_back_buffer_blur` function uses `get_blur_size` which would make it difficult to move it into render.c